<a href="https://colab.research.google.com/github/ninacchagas/pandas-exercises/blob/main/Pandas_Exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas Project Exercise 

# The Data

This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things.

All personally identifying information has been removed from the data.

Acknowledgements
The data is originally from the article Hotel Booking Demand Datasets, written by Nuno Antonio, Ana Almeida, and Luis Nunes for Data in Brief, Volume 22, February 2019.


----------------------------

#### NOTE: Names, Emails, Phone Numbers, and Credit Card numbers in the data are synthetic and not real information from people. The hotel data is real from the publication listed above.

## <div style="text-align: center">Data Column Reference</div>

<table><thead><tr class="rowsep-1"><th scope="col"><strong>Variable</strong></th><th scope="col"><strong>Type</strong></th><th scope="col"><strong>Description</strong></th><th scope="col"><strong>Source/Engineering</strong></th></tr></thead><tbody><tr><th scope="row"><em>ADR</em></th><td>Numeric</td><td>Average Daily Rate as defined by <a name="bbib5" href="#bib5" class="workspace-trigger">[5]</a></td><td>BO, BL and TR / Calculated by dividing the sum of all lodging transactions by the total number of staying nights</td></tr><tr><th scope="row"><em>Adults</em></th><td>Integer</td><td>Number of adults</td><td>BO and BL</td></tr><tr><th scope="row"><em>Agent</em></th><td>Categorical</td><td>ID of the travel agency that made the booking<a name="btbl1fna" href="#tbl1fna" class="workspace-trigger"><sup>a</sup></a></td><td>BO and BL</td></tr><tr><th scope="row"><em>ArrivalDateDayOfMonth</em></th><td>Integer</td><td>Day of the month of the arrival date</td><td>BO and BL</td></tr><tr><th scope="row"><em>ArrivalDateMonth</em></th><td>Categorical</td><td>Month of arrival date with 12 categories: “January” to “December”</td><td>BO and BL</td></tr><tr><th scope="row"><em>ArrivalDateWeekNumber</em></th><td>Integer</td><td>Week number of the arrival date</td><td>BO and BL</td></tr><tr><th scope="row"><em>ArrivalDateYear</em></th><td>Integer</td><td>Year of arrival date</td><td>BO and BL</td></tr><tr><th scope="row"><em>AssignedRoomType</em></th><td>Categorical</td><td>Code for the type of room assigned to the booking. Sometimes the assigned room type differs from the reserved room type due to hotel operation reasons (e.g. overbooking) or by customer request. Code is presented instead of designation for anonymity reasons</td><td>BO and BL</td></tr><tr><th scope="row"><em>Babies</em></th><td>Integer</td><td>Number of babies</td><td>BO and BL</td></tr><tr><th scope="row"><em>BookingChanges</em></th><td>Integer</td><td>Number of changes/amendments made to the booking from the moment the booking was entered on the PMS until the moment of check-in or cancellation</td><td>BO and BL/Calculated by adding the number of unique iterations that change some of the booking attributes, namely: persons, arrival date, nights, reserved room type or meal</td></tr><tr><th scope="row"><em>Children</em></th><td>Integer</td><td>Number of children</td><td>BO and BL/Sum of both payable and non-payable children</td></tr><tr><th scope="row"><em>Company</em></th><td>Categorical</td><td>ID of the company/entity that made the booking or responsible for paying the booking. ID is presented instead of designation for anonymity reasons</td><td>BO and BL.</td></tr><tr><th scope="row"><em>Country</em></th><td>Categorical</td><td>Country of origin. Categories are represented in the ISO 3155–3:2013 format <a name="bbib6" href="#bib6" class="workspace-trigger">[6]</a></td><td>BO, BL and NT</td></tr><tr><th scope="row"><br></th><td><br></td><td><br></td><td><br></td></tr><tr><th scope="row" rowspan="5"><em>CustomerType</em></th><td rowspan="5">Categorical</td><td>Type of booking, assuming one of four categories:</td><td rowspan="5">BO and BL</td></tr><tr><td>Contract - when the booking has an allotment or other type of contract associated to it;</td></tr><tr><td>Group – when the booking is associated to a group;</td></tr><tr><td>Transient – when the booking is not part of a group or contract, and is not associated to other transient booking;</td></tr><tr><td>Transient-party – when the booking is transient, but is associated to at least other transient booking</td></tr><tr><th scope="row"><em>DaysInWaitingList</em></th><td>Integer</td><td>Number of days the booking was in the waiting list before it was confirmed to the customer</td><td>BO/Calculated by subtracting the date the booking was confirmed to the customer from the date the booking entered on the PMS</td></tr><tr><th scope="row"><br></th><td><br></td><td><br></td><td><br></td></tr><tr><th scope="row" rowspan="7"><em>DepositType</em></th><td rowspan="7">Categorical</td><td>Indication on if the customer made a deposit to guarantee the booking. This variable can assume three categories:</td><td rowspan="2">BO and TR/Value calculated based on the payments identified for the booking in the transaction (TR) table before the booking׳s arrival or cancellation date.</td></tr><tr><td rowspan="3">No Deposit – no deposit was made;</td></tr><tr><td>In case no payments were found the value is “No Deposit”.</td></tr><tr><td rowspan="2">If the payment was equal or exceeded the total cost of stay, the value is set as “Non Refund”.</td></tr><tr><td rowspan="2">Non Refund – a deposit was made in the value of the total stay cost;</td></tr><tr><td rowspan="2">Otherwise the value is set as “Refundable”</td></tr><tr><td>Refundable – a deposit was made with a value under the total cost of stay.</td></tr><tr><th scope="row"><em>DistributionChannel</em></th><td>Categorical</td><td>Booking distribution channel. The term “TA” means “Travel Agents” and “TO” means “Tour Operators”</td><td>BO, BL and DC</td></tr><tr><th scope="row"><em>IsCanceled</em></th><td>Categorical</td><td>Value indicating if the booking was canceled (1) or not (0)</td><td>BO</td></tr><tr><th scope="row"><em>IsRepeatedGuest</em></th><td>Categorical</td><td>Value indicating if the booking name was from a repeated guest (1) or not (0)</td><td>BO, BL and C/ Variable created by verifying if a profile was associated with the booking customer. If so, and if the customer profile creation date was prior to the creation date for the booking on the PMS database it was assumed the booking was from a repeated guest</td></tr><tr><th scope="row"><em>LeadTime</em></th><td>Integer</td><td>Number of days that elapsed between the entering date of the booking into the PMS and the arrival date</td><td>BO and BL/ Subtraction of the entering date from the arrival date</td></tr><tr><th scope="row"><em>MarketSegment</em></th><td>Categorical</td><td>Market segment designation. In categories, the term “TA” means “Travel Agents” and “TO” means “Tour Operators”</td><td>BO, BL and MS</td></tr><tr><th scope="row"><br></th><td><br></td><td><br></td><td><br></td></tr><tr><th scope="row" rowspan="5"><em>Meal</em></th><td rowspan="5">Categorical</td><td>Type of meal booked. Categories are presented in standard hospitality meal packages:</td><td rowspan="5">BO, BL and ML</td></tr><tr><td>Undefined/SC – no meal package;</td></tr><tr><td>BB – Bed &amp; Breakfast;</td></tr><tr><td>HB – Half board (breakfast and one other meal – usually dinner);</td></tr><tr><td>FB – Full board (breakfast, lunch and dinner)</td></tr><tr><th scope="row"><em>PreviousBookingsNotCanceled</em></th><td>Integer</td><td>Number of previous bookings not cancelled by the customer prior to the current booking</td><td>BO and BL / In case there was no customer profile associated with the booking, the value is set to 0. Otherwise, the value is the number of bookings with the same customer profile created before the current booking and not canceled.</td></tr><tr><th scope="row"><em>PreviousCancellations</em></th><td>Integer</td><td>Number of previous bookings that were cancelled by the customer prior to the current booking</td><td>BO and BL/ In case there was no customer profile associated with the booking, the value is set to 0. Otherwise, the value is the number of bookings with the same customer profile created before the current booking and canceled.</td></tr><tr><th scope="row"><em>RequiredCardParkingSpaces</em></th><td>Integer</td><td>Number of car parking spaces required by the customer</td><td>BO and BL</td></tr><tr><th scope="row"><br></th><td><br></td><td><br></td><td><br></td></tr><tr><th scope="row" rowspan="4"><em>ReservationStatus</em></th><td rowspan="4">Categorical</td><td>Reservation last status, assuming one of three categories:</td><td rowspan="4">BO</td></tr><tr><td>Canceled – booking was canceled by the customer;</td></tr><tr><td>Check-Out – customer has checked in but already departed;</td></tr><tr><td>No-Show – customer did not check-in and did inform the hotel of the reason why</td></tr><tr><th scope="row"><em>ReservationStatusDate</em></th><td>Date</td><td>Date at which the last status was set. This variable can be used in conjunction with the <em>ReservationStatus</em> to understand when was the booking canceled or when did the customer checked-out of the hotel</td><td>BO</td></tr><tr><th scope="row"><em>ReservedRoomType</em></th><td>Categorical</td><td>Code of room type reserved. Code is presented instead of designation for anonymity reasons</td><td>BO and BL</td></tr><tr><th scope="row"><em>StaysInWeekendNights</em></th><td>Integer</td><td>Number of weekend nights (Saturday or Sunday) the guest stayed or booked to stay at the hotel</td><td>BO and BL/ Calculated by counting the number of weekend nights from the total number of nights</td></tr><tr><th scope="row"><em>StaysInWeekNights</em></th><td>Integer</td><td>Number of week nights (Monday to Friday) the guest stayed or booked to stay at the hotel</td><td>BO and BL/Calculated by counting the number of week nights from the total number of nights</td></tr><tr><th scope="row"><em>TotalOfSpecialRequests</em></th><td>Integer</td><td>Number of special requests made by the customer (e.g. twin bed or high floor)</td><td>BO and BL/Sum of all special requests</td></tr></tbody></table>

-----------

# TASKS

**Complete the tasks shown in bold below. The expected output is shown in a cell below. Be careful not to run the cell above the expected output, as it will clear the expected output. Try your best to solve these in one line of pandas code (not every single question can be solved in one line, but many can be!) Refer to solutions notebook and video to view possible solutions. NOTE: Many tasks have multiple correct solution methods!**

-----
### TASK: Run the following code to read in the "hotel_booking_data.csv" file. Feel free to explore the file a bit before continuing with the rest of the exercise.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

In [ ]:
hotels = pd.read_csv("/content/drive/MyDrive/Python Data Science Notebooks/03-Pandas/hotel_booking_data.csv")

In [ ]:
hotels.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  customer_type  \
0                        0                     0       2  ...      Transient   
1                        0                     0       2  ...      Transient   
2                        0                     1       1  ...      Transient   
3                        0                     1       1  ...      Transient   
4                        0                     2       2  ...      Transient   

    adr required_car_parking_spaces total_of_special_requests  \
0   0.0                           0                         0   
1   0.0                           0                         0   
2  75.0                           0                         0   
3  75.0                           0                         0   
4  98.0                           0                         1   

  reservation_status reservation_status_date            name  \
0          Check-Out              2015-07-01   Ernest Barnes   
1          Check-Out              2015-07-01    Andrea Baker   
2          Check-Out              2015-07-02  Rebecca Parker   
3          Check-Out              2015-07-02    Laura Murray   
4          Check-Out              2015-07-03     Linda Hines   

                         email  phone-number       credit_card  
0  Ernest.Barnes31@outlook.com  669-792-1661  ************4322  
1       Andrea_Baker94@aol.com  858-637-6955  ************9157  
2   Rebecca_Parker@comcast.net  652-885-2745  ************3734  
3            Laura_M@gmail.com  364-656-8427  ************5677  
4           LHines@verizon.com  713-226-5883  ************5498  

[5 rows x 36 columns]

---
**TASK: How many rows are there?**

In [ ]:
# CODE HERE

In [ ]:
len(hotels)

119390

There are 119390 rows

**TASK: Is there any missing data? If so, which column has the most missing data?**

In [ ]:
# CODE HERE

In [ ]:
hotels.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

There are missing data, and the company column has the most missing data.

**TASK: Drop the "company" column from the dataset.**

In [ ]:
# CODE HERE

In [ ]:
hotels = hotels.drop('company', axis=1)

**TASK: What are the top 5 most common country codes in the dataset?**

In [ ]:
# CODE HERE

In [ ]:
hotels['country'].value_counts()[:5]

PRT    48590
GBR    12129
FRA    10415
ESP     8568
DEU     7287
Name: country, dtype: int64

**TASK: What is the name of the person who paid the highest ADR (average daily rate)? How much was their ADR?**

In [ ]:
# CODE HERE

In [ ]:
hotels.sort_values('adr', ascending=False)[['adr', 'name']].iloc[0]


adr            5400.0
name    Daniel Walter
Name: 48515, dtype: object

**TASK: The adr is the average daily rate for a person's stay at the hotel. What is the mean adr across all the hotel stays in the dataset?**

In [ ]:
# CODE HERE

In [ ]:
round(hotels['adr'].mean(), 2)

101.83

**TASK: What is the average (mean) number of nights for a stay across the entire data set? Feel free to round this to 2 decimal points.**

In [ ]:
# CODE HERE

In [ ]:
hotels['total_nights'] = hotels['stays_in_week_nights'] + hotels['stays_in_weekend_nights']

In [ ]:
round(hotels['total_nights'].mean(), 2)

3.43

**TASK: What is the average total cost for a stay in the dataset? Not *average daily cost*, but *total* stay cost. (You will need to calculate total cost your self by using ADR and week day and weeknight stays). Feel free to round this to 2 decimal points.**

In [ ]:
# CODE HERE

In [ ]:
hotels['total_spend'] = hotels['adr'] * hotels['total_nights']

In [ ]:
round(hotels['total_spend'].mean(), 2)

357.85

**TASK: What are the names and emails of people who made exactly 5 "Special Requests"?**

In [ ]:
# CODE HERE

In [ ]:
hotels[hotels['total_of_special_requests'] == 5][['name', 'email']]

name                          email
7860             Amanda Harper           Amanda.H66@yahoo.com
11125            Laura Sanders      Sanders_Laura@hotmail.com
14596              Tommy Ortiz            Tommy_O@hotmail.com
14921           Gilbert Miller         Miller.Gilbert@aol.com
14922           Timothy Torres         TTorres@protonmail.com
24630          Jennifer Weaver             Jennifer_W@aol.com
27288           Crystal Horton             Crystal.H@mail.com
27477           Brittney Burke       Burke_Brittney16@att.com
29906          Cynthia Cabrera    Cabrera.Cynthia@xfinity.com
29949              Sarah Floyd              Sarah_F@gmail.com
32267           Michelle Villa         Michelle.Villa@aol.com
39027           Nichole Hebert       Hebert.Nichole@gmail.com
39129         Lindsey Mckenzie       Lindsey.Mckenzie@att.com
39525           Ashley Edwards       Edwards.Ashley@yahoo.com
70114       Christopher Torres   Torres.Christopher@gmail.com
78819   Mrs. Tara Sullivan DVM           Mrs..DVM@xfinity.com
78820           Michaela Brown          MichaelaBrown@att.com
78822        Kurt Maldonado MD              KMD15@xfinity.com
97072         Jason Richardson               Jason.R@zoho.com
97099             Terri Hurley            THurley@xfinity.com
97261        Mrs. Caitlin Webb             Mrs._W@comcast.net
98410             Holly Arroyo          Arroyo_Holly@mail.com
98674          Denise Campbell             Denise_C@gmail.com
99887            Michael Smith            Michael.S42@aol.com
99888       Dr. Trevor Sellers                  Dr._S@aol.com
101569            Kayla Murphy         Kayla.Murphy@yahoo.com
102061         Taylor Martinez    Taylor.Martinez@hotmail.com
109511          Charles Wilson       Charles_Wilson@yahoo.com
109590           Tyler Allison         Tyler.A@protonmail.com
110082          Matthew Bailey         Matthew_Bailey@aol.com
110083       Charlotte Acevedo        Charlotte_A@verizon.com
111909         Darrell Brennan  Brennan_Darrell51@hotmail.com
111911          Melinda Jensen         MelindaJensen@zoho.com
113915            Terry Arnold          Arnold.Terry@zoho.com
114770             Mary Nguyen     Nguyen.Mary@protonmail.com
114909          Lindsay Cuevas      Lindsay.Cuevas40@mail.com
116455       Cynthia Hernandez   CynthiaHernandez@xfinity.com
116457          Angela Hawkins             Angela_H@gmail.com
118817              Sue Lawson            Sue.L52@comcast.net
119161         Alyssa Richards        Alyssa_Richards@aol.com

**TASK: What percentage of hotel stays were classified as "repeat guests"? (Do not base this off the name of the person, but instead of the is_repeated_guest column)**

In [ ]:
#CODE HERE

In [ ]:
hotels['is_repeated_guest'].unique()

array([0, 1])

In [ ]:
round(100 * sum(hotels['is_repeated_guest'] == 1) / len(hotels), 2)

3.19

**TASK: What are the top 5 most common last name in the dataset? Bonus: Can you figure this out in one line of pandas code? (For simplicity treat the a title such as MD as a last name, for example Caroline Conley MD can be said to have the last name MD)**

In [ ]:
#CODE HERE

In [ ]:
hotels['name'].apply(lambda name: name.split()[-1]).value_counts()[: 5]

Smith       2503
Johnson     1990
Williams    1618
Jones       1434
Brown       1423
Name: name, dtype: int64

**TASK: What are the names of the people who had booked the most number children and babies for their stay? (Don't worry if they canceled, only consider number of people reported at the time of their reservation)**

In [ ]:
#CODE HERE

In [ ]:
hotels['total_of_kids'] = hotels['children'] + hotels['babies']

In [ ]:
hotels.sort_values('total_of_kids', ascending=False)[['total_of_kids', 'name']]

total_of_kids               name
328              10.0      Jamie Ramirez
46619            10.0    Nicholas Parker
78656             9.0      Marc Robinson
19718             3.0  Mr. Jeffrey Cross
107837            3.0      Albert French
...               ...                ...
119389            0.0     Ariana Michael
40600             NaN       Craig Campos
40667             NaN       David Murphy
40679             NaN       Frank Burton
41160             NaN      Jerry Roberts

[119390 rows x 2 columns]

**TASK: What are the top 3 most common area code in the phone numbers? (Area code is first 3 digits)**

In [ ]:
#CODE HERE

In [ ]:
hotels['phone-number'].apply(lambda num: num[:3]).value_counts()[:3]

799    168
185    167
541    166
Name: phone-number, dtype: int64

**TASK: How many arrivals took place between the 1st and the 15th of the month (inclusive of 1 and 15) ? Bonus: Can you do this in one line of pandas code?**

In [ ]:
#CODE HERE

In [ ]:
hotels['arrival_date_day_of_month'].apply(lambda day: day in range(1, 16)).sum()

58152